test/cipher_test.rb

```diff
@@ -22,4 +21,54 @@ def test_it_can_create_a_character_set
   def test_it_can_return_todays_date
     assert_equal '040895', @cipher.date
   end
+
+  def test_it_can_generate_key
+    @cipher.stubs(:rand).returns('2715')
+    assert_equal '02715', @cipher.generate_key
+  end
+
+  def test_it_can_square_date
+    assert_equal 1672401025, @cipher.squared_date('040895')
+  end
+
+  def test_it_can_generate_offsets
+    expected_offsets = {a: 1, b: 0, c: 2, d: 5}
+    assert_equal expected_offsets, @cipher.generate_offsets('040895')
+  end
+
+  def test_it_can_separate_keys
+    expected_keys = {a: 02, b: 27, c: 71, d: 15}
+    assert_equal expected_keys, @cipher.separate_keys('02715')
+  end
+
+  def test_it_can_generate_shifts
+    expected_shifts = {a: 3, b: 27, c: 73, d: 20}
+    assert_equal expected_shifts, @cipher.generate_shifts('02715', '040895', 1)
+
+    expected_shifts2 = {:a=>-3, :b=>-27, :c=>-73, :d=>-20}
+    assert_equal expected_shifts2, @cipher.generate_shifts('02715', '040895', -1)
+  end
+
+  def test_it_can_shift_character_set
+    shifts = {a: 3, b: 27, c: 73, d: 20}
+    expected_set = ['u', 'v', 'w', 'x', 'y', 'z', ' ', 'a', 'b', 'c', 'd', 'e', 'f',
+                    'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't']
+    assert_equal expected_set, @cipher.shift_char_set(3, shifts)
+
+
+    expected_set2 = ['t', 'u', 'v', 'w', 'x', 'y', 'z', ' ', 'a', 'b', 'c', 'd', 'e',
+                    'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's']
+    assert_equal expected_set2, @cipher.shift_char_set(14, shifts)
+  end
+
+  def test_it_can_mutate_message
```

Comment:
I was debating whether this method should live in Cipher or Enigma, but because it's solely mutating the char_set I decided to move it to Cipher.